nav2_tasks/include/nav2_tasks/stop_action.hpp

```diff
@@ -0,0 +1,142 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_TASKS__STOP_ACTION_HPP_
+#define NAV2_TASKS__STOP_ACTION_HPP_
+
+#include <string>
+#include <memory>
+#include <cmath>
+#include <thread>
+#include <chrono>
+
+#include "rclcpp/rclcpp.hpp"
+#include "geometry_msgs/msg/twist.hpp"
+#include "nav2_tasks/follow_path_task.hpp"
+#include "nav2_robot/robot.hpp"
+
+using namespace std::chrono_literals; // NOLINT
+
+namespace nav2_tasks
+{
+
+class StopAction : public BT::ActionNode
+{
+public:
+  explicit StopAction(const std::string & action_name)
+  : BT::ActionNode(action_name), initialized_(false)
+  {
+  }
+
+  StopAction() = delete;
+
+  ~StopAction() {}
+
+  BT::NodeStatus tick() override
+  {
+    // A BT node can't get values from the blackboard in the constructor since
+    // the BT library doesn't set the blackboard until after the tree if build.
+    if (!initialized_) {
+    node_ = blackboard()->template get<rclcpp::Node::SharedPtr>("node");
+
+    robot_ = std::make_unique<nav2_robot::Robot>(node_);
+
+    controller_client_ = std::make_unique<nav2_tasks::FollowPathTaskClient>(node_);
+
+    initialized_ = true;
+    }
+
+    static auto start_time = std::chrono::system_clock::now();
+
+    static bool new_call = true;
+
+    if (new_call) {
+      RCLCPP_INFO(node_->get_logger(), "StopAction: cancelling path following task server");
+      controller_client_->cancel();
+
+      RCLCPP_INFO(node_->get_logger(), "StopAction: publishing zero velocity command");
+
+      // TODO(orduno) Implement a stop method on the robot
+      // robot_.stop();
+
+      geometry_msgs::msg::Twist twist;
+      twist.linear.x = 0.0;
+      twist.linear.y = 0.0;
+      twist.linear.z = 0.0;
+      twist.angular.x = 0.0;
+      twist.angular.y = 0.0;
+      twist.angular.z = 0.0;
+      robot_->sendVelocity(twist);
+
+      start_time = std::chrono::system_clock::now();
+      new_call = false;
+      RCLCPP_INFO(node_->get_logger(), "StopAction: stabilizing robot");
+
+      // return BT::NodeStatus::RUNNING;
+    }
+
+    // TODO(orduno) What should this node return for a long running action
+
+    // Give some time to the robot to get stable
+    // TODO(orduno) Use the robot interface instead
+    // robot_.waitUntilStable();
+
+    // auto current_time = std::chrono::system_clock::now();
+    // std::chrono::duration<double> elapsed_time = current_time - start_time;
+    // std::chrono::duration<double> hold_still_time(5);
+    // std::chrono::duration<double> remaining = hold_still_time - elapsed_time;
+
+    // if (elapsed_time < hold_still_time) {
+    //   RCLCPP_INFO(get_logger(), "tick:: %.2f seconds remaining.",
+    //     remaining.count());
+    //   return BT::NodeStatus::RUNNING;
+    // }
+
+    // RCLCPP_INFO(get_logger(), "tick:: completed");
+    // new_call = true;
+
+    auto sleep_time = 5s;
```

Comment:
Sleeping in a tick function will pause the whole tree. Instead of sleeping you can monitor the elapsed time.

Reply:
Changed with the new refactored code.